### DISCUSSION TITLE
Energy usage

### DISCUSSION THREAD

**Discussion by RonnyRusten** [robbinjanssen/python-ojmicroline-thermostat#104] [General]
I am trying to get the energy usage from the thermostat, as we discussed in the HA integration repo.
It's my first attempt at contributing on GH, so I have some questions.
1. Should I open an issue, or should I just make a PR if I ever get something sensible out of this?
2. Is there any naming standards for branches I should follow?

**Comment by robbinjanssen** [comment]
Hi, thanks for attempting! :-)

I've created an issue, see https://github.com/robbinjanssen/python-ojmicroline-thermostat/issues/105
Don't worry about the branch names. Personally I always do something like `robbinjanssen/add-energy`.

**Reply by RonnyRusten** [reply]
Great, I'll probably ask more newbie questions when/if I get as far as to a PR :-)

**Comment by RonnyRusten** [comment]
I'm having trouble getting info from the EnergyUsage api. All the values are 0, for all ViewTypes. Is this just me, or is it like this for all? I use "yesterday" as the date, I guess today isn't updated until tomorrow at 00:00:00 (or maybe a little after that? My app is showing (some strange) values, I don't know where it get its values from, maybe I need to try to sniff the traffic?

**Reply by robbinjanssen** [reply]
I've done some quick sniffing in the past, and found out that doing something like this gives results for a year:

```python
data = await self._request(
    "api/EnergyUsage/GetEnergyUsage",
    method=hdrs.METH_POST,
    params={
        "sessionid": self.__session_id,
    },
    body={
        "APIKEY": self.__api_key,
        "History": 0,
        "ThermostatID": thermostat.serial_number,
        "ViewType": 4,
        "History": 0,
        "DateTime": "2022-10-11T15:29:35"
    }
)
```

**Reply by RonnyRusten** [reply]
Ok, so you use the serial number instead of the id? That's odd, but I'll test it later today.

**Reply by RonnyRusten** [reply]
That actually worked, thanks.

**Reply by robbinjanssen** [reply]
Yes I found that it is using the serial number as ID while sniffing, it is quite confusing indeed 😂

**Comment by RonnyRusten** [comment]
OK, new newbie question.
I have made some changes to my fork of this repo, and would like to test in HA. (https://github.com/RonnyRusten/python-ojmicroline-thermostat/tree/rr-energy-usage) (Don't laugh! :-))
You said I could modify the manifest of the integration, but, how? I read here (https://developers.home-assistant.io/docs/creating_integration_manifest/#custom-requirements-during-development--testing) that I can set it to my repo and branch, but only if version x is not already installed. The question is then, how do I change the version of my library? I guess it will not load my library because it already has your library installed... Do I uninstall your library? (If yes, how?)
As I said, I'm a totally newbie :-)

**Reply by robbinjanssen** [reply]
I have to say that was quite a bit confusing for me as well.. What I did was update the [version](https://github.com/RonnyRusten/python-ojmicroline-thermostat/blob/rr-energy-usage/pyproject.toml#L3) with a little bump, for example v1.0.1 and then change the manifest.json in your custom component folder to something like this:

```
    "requirements": [
        "git+https://github.com/RonnyRusten/python-ojmicroline-thermostat.git@rr-energy-usage#ojmicroline-thermostat==1.0.1"
    ],
```

Whenever I made another changed I bumped the version to 1.0.2 etc. etc.

If there's a better way please let me know 😂

**Reply by RonnyRusten** [reply]
If you have done this and it works, it's good enough for me😊

**Comment by TorStava** [comment]
Thank you @robbinjanssen for your excellent work on this library. 

I was getting ready to replace my two Microtemp thermostats due to missing implementation in Home Assistant. You saved me the trouble ;)

I have implemented the code for fetching the energy consumption in pull request #468. As discussed there I think the discrepancy you have seen between the API numbers and the app report, is due to the Energy Tariff option on the thermostat. This needs to be disabled on the thermostat for the numbers to be reported as expected. I have included a note about this in the README.

I have also proposed a pull request for the custom component for Home Asssistant that will utilise this new data from the Python library.

**Comment by TorStava** [comment]
I've been using the implementation (ref. pull request #468) for a few days now and have found an annoying bug. 
Since the API only seem to report energy consumption about once per hour, the first hour of the day always reports 0 kWh, and the second hour will report the actual usage for the first hour. I will take a look how this can be resolved, possbily by using monthly values and calculating deltas instead of using the reported value directly. Any good ideas are welcome.

![image](https://github.com/user-attachments/assets/512b2bb2-6e71-4717-b160-5dc83f9b0afd)

![image](https://github.com/user-attachments/assets/e962f17c-6628-486b-94ec-5392d44dec12)
